robowflex_library/src/robot.cpp

```diff
@@ -579,84 +556,150 @@ bool Robot::hasJoint(const std::string &joint) const
+  : IKQuery(group,                         //
+            Geometry::makeSphere(radius),  //
+            TF::createPoseXYZ(position),   //
+            orientation,                   //
+            tolerance)
+{
+}
 
-    return false;
+Robot::IKQuery::IKQuery(const std::string &group, const GeometryConstPtr &region, const RobotPose &pose,
+                        const Eigen::Quaterniond &orientation, const Eigen::Vector3d &tolerance,
+                        const ScenePtr &scene, bool verbose)
+  : group(group), scene(scene), verbose(scene and verbose)
+{
+    addRequest("", region, pose, orientation, tolerance);
+}
+
+Robot::IKQuery::IKQuery(const std::string &group, const RobotPoseVector &poses,
+                        const std::vector<std::string> &input_tips, double radius,
+                        const Eigen::Vector3d &tolerance, const ScenePtr &scene, bool verbose)
+  : group(group), scene(scene), verbose(scene and verbose)
+{
+    if (poses.size() != input_tips.size())
+        throw std::runtime_error("Invalid multi-target IK query. poses != tips.");
+
+    for (std::size_t i = 0; i < poses.size(); ++i)
+        addRequest(input_tips[i],                              //
+                   Geometry::makeSphere(radius),               //
+                   TF::createPoseXYZ(poses[i].translation()),  //
+                   Eigen::Quaterniond{poses[i].rotation()},    //
+                   tolerance);
+}
+
+void Robot::IKQuery::addRequest(const std::string &tip, const GeometryConstPtr &region, const RobotPose &pose,
+                                const Eigen::Quaterniond &orientation, const Eigen::Vector3d &tolerance)
+{
+    tips.emplace_back(tip);
+    regions.emplace_back(region);
+    region_poses.emplace_back(pose);
+    orientations.emplace_back(orientation);
+    tolerances.emplace_back(tolerance);
 }
 
-bool Robot::setFromIKCollisionAware(const ScenePtr &scene, const std::string &group,
-                                    const GeometryConstPtr &region, const RobotPose &pose,
-                                    const Eigen::Quaterniond &orientation, const Eigen::Vector3d &tolerances,
-                                    bool verbose)
+bool Robot::IKQuery::sampleRegions(RobotPoseVector &poses) const
 {
-    robot_model::JointModelGroup *jmg = model_->getJointModelGroup(group);
+    const std::size_t n = regions.size();
+    poses.resize(n);
 
-    const auto &gsvcf =                                               //
-        [&scene, &verbose](robot_state::RobotState *state,            //
-                           const moveit::core::JointModelGroup *jmg,  //
-                           const double *values)                      //
+    bool sampled = true;
+    for (std::size_t j = 0; j < n and sampled; ++j)
     {
-        state->setJointGroupPositions(jmg, values);
-        state->updateCollisionBodyTransforms();
+        const auto &point = regions[j]->sample();
+        if ((sampled &= point.first))
```

Comment:
Is double parenthesis needed here?

Reply:
yeah, the double is needed so it actually sets `sampled`